File: assign2-DewsDavid31.swift
// Assign 02, translated into swift to the best of my ability
import Foundation

assign02_main()
typealias split3Ways = (first: [Int], second: [Int], third: [Int])
typealias tuple = (list: [Int], time: Float)
typealias split2Ways = (first: [Int], second: [Int])


func bubbleSort(listOfItems: [Int]) -> tuple {
	var tempList = listOfItems
	let startTime = DispatchTime.now().uptimeNanoseconds
	for offset in 1..<listOfItems.count {
		for index in 0..<(listOfItems.count - offset) {
			if(tempList[index] > tempList[index + 1]){
				let temp = tempList[index]
				tempList[index] = tempList[index + 1]
				tempList[index + 1] = temp
			}
		}
	}
	let endTime = DispatchTime.now().uptimeNanoseconds
	return tuple(list: tempList, time: ((Float(endTime) - Float(startTime))/1000000000))
}

func mergeSort(listOfItems: [Int], splitBy3: Bool) -> tuple{
	let startTime = DispatchTime.now().uptimeNanoseconds
	let nextList = listOfItems
	let tempList = mergeSortRecursive(listOfItems: nextList,splitBy3: splitBy3)
	let endTime = DispatchTime.now().uptimeNanoseconds
	return tuple(list: tempList, time: ((Float(endTime) - Float(startTime))/1000000000))
}

func mergeSortRecursive(listOfItems: [Int], splitBy3: Bool) -> [Int]{
	if(listOfItems.count <= 1){
		return listOfItems
	}
	else if(splitBy3){
		let splittedList = splitTo3(listOfItems: listOfItems)
		return merge(left: merge(left: mergeSortRecursive(listOfItems: splittedList.first,splitBy3: splitBy3),right:  mergeSortRecursive(listOfItems: splittedList.second,splitBy3: splitBy3)),right: mergeSortRecursive(listOfItems: splittedList.third,splitBy3: splitBy3))
	}
	let splittedList = splitTo2(listOfItems: listOfItems)
	return merge(left: mergeSortRecursive(listOfItems: splittedList.first,splitBy3: splitBy3),right: mergeSortRecursive(listOfItems: splittedList.second,splitBy3: splitBy3))
}

func merge(left: [Int], right: [Int]) -> [Int] {
	var temp = left + right
	var leftIndex = 0
	var rightIndex = 0
	var tempIndex = 0
	if(temp.count <= 1){
		return temp
	}
	while(leftIndex < left.count && rightIndex < right.count){
		let leftItem = left[leftIndex]
		let rightItem = right[rightIndex]
		if(leftItem < rightItem) {
			temp[tempIndex] = leftItem
			leftIndex += 1
			tempIndex += 1
		}
		else{
			temp[tempIndex] = rightItem
			rightIndex += 1
			tempIndex += 1
		}
	}
	while(leftIndex < left.count){
		temp[tempIndex] = left[leftIndex]
		tempIndex += 1
		leftIndex += 1
	}
	while(rightIndex < right.count ){
		temp[tempIndex] = right[rightIndex]
		tempIndex += 1
		rightIndex += 1
	}
	return temp
}

func splitTo2(listOfItems: Array<Int>) -> split2Ways {
	let halfway = Int(ceil(Double(listOfItems.count) / Double(2)))
	return split2Ways(first: Array<Int>(listOfItems[0..<halfway]), second: Array<Int>(listOfItems[halfway..<listOfItems.count]))
}

func splitTo3(listOfItems: Array<Int>) -> split3Ways {
	let third = Int(ceil(Double(listOfItems.count) / Double(3)))
	let double = third * 2
	return split3Ways(first: Array<Int>(listOfItems[0..<third]), second: Array<Int>(listOfItems[third..<double]), third: Array<Int>(listOfItems[double..<listOfItems.count]))
}


func quickSort(listOfItems: [Int], pivotToUse: String) -> tuple {
	let startTime = DispatchTime.now().uptimeNanoseconds
	let endTime = DispatchTime.now().uptimeNanoseconds
	let tempList = quickSortRecursive(listOfItems: listOfItems, pivotToUse: pivotToUse)
	return tuple(list: tempList, time: ((Float(endTime) - Float(startTime))/1000000000))
}

func quickSortRecursive(listOfItems: [Int], pivotToUse: String) -> [Int] {
	if(listOfItems.count <= 1){
		return listOfItems
	}
	var pivot = 0
	if(pivotToUse == "first"){
		pivot = Int(ceil(Double(listOfItems.count))/Double(2))
	}
	var right: [Int] = []
	var left: [Int] = []
	let pivotVal = listOfItems[pivot]
	for item in listOfItems{
		if(item > pivotVal){
			right.append(item)
		}
		else if(item < pivotVal){
			left.append(item)
		}
	}
	return quickSortRecursive(listOfItems: Array<Int>(left), pivotToUse: pivotToUse) + [pivotVal] + quickSortRecursive(listOfItems: Array<Int>(right), pivotToUse: pivotToUse)
}

func radixSort(listOfItems: [Int], maxDigits: Int) -> tuple {
	let startTime = DispatchTime.now().uptimeNanoseconds
	let endTime = DispatchTime.now().uptimeNanoseconds
	//TODO: do me
	var tempList = listOfItems
	var base = 1
	var done = false
	while(!done){
		done = true
		var buckets: [[Int]] = .init(repeating: [], count: 10)
		for item in tempList{
			let remainingPart = item / base
			if(remainingPart > 0){
				done = false
			}
			let digit = remainingPart % 10
			buckets[digit].append(item)
		}
		base *= 10
		tempList = buckets.flatMap({$0})
	}
	return tuple(list: tempList, time: ((Float(endTime) - Float(startTime))/1000000000))
}

func assign02_main(){
	let listRange: CountableRange<Int> = 0..<5000
	var list1: [Int] = Array(listRange)
	list1.shuffle()
	


	var bubbleRes = bubbleSort(listOfItems: list1)
	var mergeRes2 = mergeSort(listOfItems: list1, splitBy3: false)
	var mergeRes3 = mergeSort(listOfItems: list1, splitBy3: true)
	var quickResA = quickSort(listOfItems: list1, pivotToUse: "first")
	var quickResB = quickSort(listOfItems: list1, pivotToUse: "middle")
	var radixRes = radixSort(listOfItems: list1, maxDigits: String(list1.max()!).count)

	var bubbleTime = String(format: "%.4f", bubbleRes.time)
	var merge2Time = String(format: "%.4f", mergeRes2.time)
	var merge3Time = String(format: "%.4f", mergeRes3.time)
	var quickATime = String(format: "%.4f", quickResA.time)
	var quickBTime = String(format: "%.4f", quickResB.time)
	var radixTime = String(format: "%.4f", radixRes.time)



	print("\nlist1 results (randomly shuffled w/ size = \(list1.count))")
	print(list1[..<10])
	print("  bubbleSort time: \(bubbleTime) sec")
	print(bubbleRes.list[..<10])
	print("  mergeSort time: \(merge2Time) sec")
	print(mergeRes2.list[..<10])
	print("  mergeSort2 time: \(merge3Time) sec")
	print(mergeRes3.list[..<10])
	print("  quickSortA time: \(quickATime) sec")
	print(quickResA.list[..<10])
	print("  quickSortB time: \(quickBTime) sec")
	print(quickResB.list[..<10])
	print("  radixSort time: \(radixTime) sec")
	print(radixRes.list[..<10])


	let listRange2: CountableRange<Int> = (1000..<6000)
	let list2: [Int] = Array(listRange2.reversed())
	
	bubbleRes = bubbleSort(listOfItems: list2)
	mergeRes2 = mergeSort(listOfItems: list2, splitBy3: false)
	mergeRes3 = mergeSort(listOfItems: list2, splitBy3: true)
	quickResA = quickSort(listOfItems: list2, pivotToUse:"first")
	quickResB = quickSort(listOfItems: list2, pivotToUse: "middle")
	radixRes = radixSort(listOfItems: list2, maxDigits: String(list2.max()!).count)

	bubbleTime = String(format: "%.4f", bubbleRes.time)
	merge2Time = String(format: "%.4f", mergeRes2.time)
	merge3Time = String(format: "%.4f", mergeRes3.time)
	quickATime = String(format: "%.4f", quickResA.time)
	quickBTime = String(format: "%.4f", quickResB.time)
	radixTime = String(format: "%.4f", radixRes.time)



	print("\nlist2 results (sorted in reverse w/ size = \(list2.count))")
	print(list2[..<10])
	print("  bubbleSort time: \(bubbleTime) sec")
	print(bubbleRes.list[..<10])
	print("  mergeSort time: \(merge2Time) sec")
	print(mergeRes2.list[..<10])
	print("  mergeSort2 time: \(merge3Time) sec")
	print(mergeRes3.list[..<10])
	print("  quickSortA time: \(quickATime) sec")
	print(quickResA.list[..<10])
	print("  quickSortB time: \(quickBTime) sec")
	print(quickResB.list[..<10])
	print("  radixSort time: \(radixTime) sec")
	print(radixRes.list[..<10])

}

File: assign2-DewsDavid31.py
"""
Assign 02 - David Dews

Directions:
    * Complete the sorting algorithm functions that are given below. Note that
      it okay (and probably helpful) to define auxiliary/helper functions that
      are called from the functions below.  Refer to the README.md file for
      additional info.

    * NOTE: Remember to add a docstring for each function, and that a reasonable
      coding style is followed (e.g. blank lines between functions).
      Your program will not pass the tests if this is not done!

    * Be sure that you implement your own sorting functions since.
      No credit will be given if Python's built-in sort function is used.
"""

import time
import random
import sys
from math import ceil, log10

# Set recursion limit to be greater than the max size of any list you attempt to sort
sys.setrecursionlimit(10000)


def bubbleSort(list_of_items):
    """ Sorts list of integers by swapping each pair across"""
    start_time = time.time()

    # INSERT YOUR BUBBLE SORT CODE HERE
    for offset in range(1, len(list_of_items)):
        for index in range(len(list_of_items) - offset):
            if list_of_items[index] > list_of_items[index + 1]:
                temp = list_of_items[index]
                list_of_items[index] = list_of_items[index + 1]
                list_of_items[index + 1] = temp

    elapsed_time = time.time() - start_time
    return (list_of_items, elapsed_time)


def mergeSort(list_of_items, split_by_3=False):
    """ Sorts by dividing list until singleton and inserting pairs"""
    start_time = time.time()

    # INSERT YOUR MERGE SORT CODE HERE...
    #   * SPLITTING BY 2 WHEN split_by_3 = False
    #   * SPLITTING BY 3 WHEN split_by_3 = True
    list_of_items = _merge_sort_recursive(list_of_items, split_by_3)
    elapsed_time = time.time() - start_time
    return (list_of_items, elapsed_time)


def _split_by_3_(list_of_items):
    """ Private helper method for splitting lists in 3"""
    third = round(len(list_of_items) / 3)
    double = third * 2
    return list_of_items[:third], list_of_items[third:double], list_of_items[double:]


def _split_by_2_(list_of_items):
    """ Private helper method for splitting lists in half"""
    halfway = round(len(list_of_items) / 2)
    return list_of_items[:halfway], list_of_items[halfway:]


def _merge_(left, right):
    """ Private helper method that merges two lists in order"""
    temp = left + right
    left_index, right_index, temp_index = 0, 0, 0
    while left_index < len(left) and right_index < len(right):
        left_item = left[left_index]
        right_item = right[right_index]
        if left_item < right_item:
            temp[temp_index] = left_item
            left_index += 1
        else:  # right is smallest
            temp[temp_index] = right_item
            right_index += 1
        temp_index += 1
    while left_index < len(left):
        temp[temp_index] = left[left_index]
        left_index += 1
        temp_index += 1
    while right_index < len(right):
        temp[temp_index] = right[right_index]
        right_index += 1
        temp_index += 1
    return temp


def _merge_sort_recursive(list, by_3):
    """ Private recursive helper function for mergeSort calls"""
    if len(list) <= 1:
        return list
    elif by_3:
        first, second, third = _split_by_3_(list)
        return (_merge_(_merge_(_merge_sort_recursive(first, by_3), _merge_sort_recursive(second, by_3)),
                        _merge_sort_recursive(third, by_3)))
    else:
        first, second = _split_by_2_(list)
        return _merge_(_merge_sort_recursive(first, by_3), _merge_sort_recursive(second, by_3))


def quickSort(list_of_items, pivot_to_use='first'):
    """ Sorts list of integers by repeatedly picking a pivot and splitting list over it"""
    start_time = time.time()
    # INSERT YOUR QUICK SORT CODE HERE...
    #  * USING FIRST ITEM IN THE LIST AS THE PIVOT WHEN pivot_to_use = 'first'
    #  * USING MIDDLE ITEM IN THE LIST AS THE PIVOT WHEN pivot_to_use != 'first'
    # AND BE SURE THAT CONTINUES FOR SUBSEQUENT/RECURSIVE CALLS AS WELL
    new_list = _quick_sort_recursive_(list_of_items, pivot_to_use)
    elapsed_time = time.time() - start_time
    return (new_list, elapsed_time)


def _quick_sort_recursive_(list_of_items, pivot_to_use):
    """ Private recursive helper method for QuickSort calls"""
    if len(list_of_items) <= 1:
        return list_of_items
    pivot = 0
    if not pivot_to_use == 'first':
        pivot = round(len(list_of_items) / 2)
    right = []
    left = []
    pivot_val = list_of_items[pivot]
    for item in list_of_items:
        if item > pivot_val:
            right.append(item)
        elif item < pivot_val:
            left.append(item)
    return _quick_sort_recursive_(left, pivot_to_use) + [pivot_val] + _quick_sort_recursive_(right, pivot_to_use)


def radixSort(list_of_items, max_digits):
    """ Sorts a list of integers by comparing chars and binning elements"""
    start_time = time.time()

    # INSERT YOUR RADIX SORT CODE HERE
    padded_list = []
    buckets = []
    resulting_list = []
    for item in list_of_items:
        padded_list.append(_pad_(item, max_digits))
    for item in padded_list:
        _place_(buckets, item)
    _traverse_(buckets, resulting_list)

    elapsed_time = time.time() - start_time
    return (resulting_list, elapsed_time)


def _place_(buckets, item):
    """ Private helper method for radixSort that upserts arrays for bins"""
    for index in range(len(item)):
        char = item[index]
        digit = int(char)
        while len(buckets) <= digit:
            buckets.append([])
        if type(buckets[digit]) != list:
            buckets[digit] = []
        if index == len(item) - 1:
            buckets[digit].append(int(item))
        buckets = buckets[digit]


def _traverse_(bucket, result):
    """ Private helper method for radixSort that traverses the list of bins in order"""
    if isinstance(bucket, int):
        result.append(bucket)
    if isinstance(bucket, list):
        for sub_bucket in bucket:
            _traverse_(sub_bucket, result)


def _pad_(item, size):
    """ Private helper method for radixSort that pads up to size with leading '0'"""
    current_size = len(str(item))
    if current_size > size:
        return str(item)
    else:
        return "0" * (size - current_size) + str(item)


def assign02_main():
    """ A 'main' function to be run when our program is run standalone """
    list1 = list(range(5000))
    random.seed(1)
    random.shuffle(list1)

    # list1 = [54, 26, 93, 17, 77, 31, 44, 55, 20] # helpful for early testing

    # run sorting functions
    bubbleRes = bubbleSort(list(list1))
    mergeRes2 = mergeSort(list(list1), split_by_3=False)
    mergeRes3 = mergeSort(list(list1), split_by_3=True)
    quickResA = quickSort(list(list1), pivot_to_use='first')
    quickResB = quickSort(list(list1), pivot_to_use='middle')
    radixRes = radixSort(list(list1), ceil(log10(max(list1))))

    # Print results
    print(f"\nlist1 results (randomly shuffled w/ size = {len(list1)})")
    print(list1[:10])
    print(f"  bubbleSort time: {bubbleRes[1]:.4f} sec")
    print(bubbleRes[0][:10])
    print(f"  mergeSort2 time: {mergeRes2[1]:.4f} sec")
    print(mergeRes2[0][:10])
    print(f"  mergeSort3 time: {mergeRes3[1]:.4f} sec")
    print(mergeRes3[0][:10])
    print(f"  quickSortA time: {quickResA[1]:.4f} sec")
    print(quickResA[0][:10])
    print(f"  quickSortB time: {quickResB[1]:.4f} sec")
    print(quickResB[0][:10])
    print(f"  radixSort time: {radixRes[1]:.4f} sec")
    print(radixRes[0][:10])

    # Try with a list sorted in reverse order (worst case for quicksort)
    list2 = list(range(6000, 1000, -1))

    # run sorting functions
    bubbleRes = bubbleSort(list(list2))
    mergeRes2 = mergeSort(list(list2), split_by_3=False)
    mergeRes3 = mergeSort(list(list2), split_by_3=True)
    quickResA = quickSort(list(list2), pivot_to_use='first')
    quickResB = quickSort(list(list2), pivot_to_use='middle')
    radixRes = radixSort(list(list2), ceil(log10(max(list2))))

    # Print results
    print(f"\nlist2 results (sorted in reverse w/ size = {len(list2)})")
    print(list2[:10])
    print(f"  bubbleSort time: {bubbleRes[1]:.4f} sec")
    print(bubbleRes[0][:10])
    print(f"  mergeSort2 time: {mergeRes2[1]:.4f} sec")
    print(mergeRes2[0][:10])
    print(f"  mergeSort3 time: {mergeRes3[1]:.4f} sec")
    print(mergeRes3[0][:10])
    print(f"  quickSortA time: {quickResA[1]:.4f} sec")
    print(quickResA[0][:10])
    print(f"  quickSortB time: {quickResB[1]:.4f} sec")
    print(quickResB[0][:10])
    print(f"  radixSort time: {radixRes[1]:.4f} sec")
    print(radixRes[0][:10])


# Check if the program is being run directly (i.e. not being imported)
if __name__ == '__main__':
    assign02_main()
